### PR TITLE
Voting for boardgames

### PR DESCRIPTION
**Pull request by zofiakochutek**
Resolves issue #20.

Do not merge yet.

### CODE REVIEW

**Comment by Bozydarek** [comment]
Hi,
I see that you have made some great progress here. Thanks for your contribution to the project! :tada:
I don't know if you plan some more changes here but I already have some comments to your PR.

- I saw that you have some unresolved conflict after merge to master (e.g. https://github.com/ksiuwr/zosia16-site/pull/83/files#diff-565e4be1ff29968e3e1661510cbf6016R7)
- I suggest you to "reset" your migrations (remove all migrations for boardgames and generate a single one)
- I don't know if you know that but you can run tests locally - just use `./tools/dev.sh test`
- Please avoid using Polish language in comments :wink: 

I haven't made a full code review yet, so probably there will be some more changes to do, but I will wait with that until you confirm that you are ready to merge that :slightly_smiling_face:

File: app/src/boardgames/views.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    return render(request, 'boardgames/index.html', ctx)
+
+
+@login_required
+@require_http_methods(['GET', 'POST'])
+def my_boardgames(request):
+    user_boardgames = Boardgame.objects.filter(user=request.user)
+    can_add = False
+    if user_boardgames.count() < 3:
+        can_add = True
```
Comment (ref-humbold) [inline app/src/boardgames/views.py]: This can be simplified to a single line :)

File: app/src/boardgames/views.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    start_index = boargamegeek_html.find(title_bytes)
+    start_index += len(title_str)
+    end_index = boargamegeek_html.find(
+        bytearray(' |', encoding), start_index)
+    name_bytes = boargamegeek_html[start_index: end_index]
+    name_str = name_bytes.decode(encoding)
+    if name_str == "BoardGameGeek":
+        pass
+        # TODO: jakiś error
+    return name_str
```
Comment (ref-humbold) [inline app/src/boardgames/views.py]: I suppose an error should be thrown here

File: app/src/boardgames/views.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    if user_boardgames.count() < 3:
+        can_add = True
+    ctx = {'user_boardgames': user_boardgames,
+           'can_add': can_add}
+    return render(request, 'boardgames/my_boardgames.html', ctx)
+
+
+# TODO: check with regex and check if not already in the database
+def validate_url(url):
+    pass
```
Comment (ref-humbold) [inline app/src/boardgames/views.py]: Either write this validation function or remove it

File: app/src/boardgames/views.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+@staff_member_required
+@require_http_methods(['POST'])
+def vote_edit(request):
+    votes = Vote.objects.filter(
+        user=request.user).values_list('boardgame', flat=True)
+    old_ids = list(votes)
+    new_ids = json.loads(request.POST.get('new_ids'))
+    if len(new_ids) > 3:
+        return Http404
```
Comment (ref-humbold) [inline app/src/boardgames/views.py]: Http404 is an exception and should be raised with an appropriate message

File: app/src/boardgames/views.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+import json
+
+from urllib.request import urlopen
+
+
+@login_required
+@require_http_methods(['GET'])
+def index(request):
+    boardgames = Boardgame.objects.all()
+    boardgames = sorted(boardgames, key=lambda x: x.votes_amount, reverse=True)
```
Comment (ref-humbold) [inline app/src/boardgames/views.py]: You may use Django QuerySet methods instead, check out other places as well

File: app/src/boardgames/views.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    for x in new_ids:
+        toggle_accepted(x)
+    return JsonResponse({'old_ids': old_ids, 'new_ids': new_ids})
+
+
+@staff_member_required
+@require_http_methods(['POST'])
+def boardgame_delete(request):
+    boardgame_id = request.POST.get('boardgame_id')
+    print(boardgame_id)
```
Comment (ref-humbold) [inline app/src/boardgames/views.py]: Why are you using print function here?

File: app/src/templates/checkboxes.html [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            var new_info = getNewInfo();
+            var new_ids = new_info[0];
+            var new_names = new_info[1];
+            console.log(new_ids, new_names);
+            changeVote(new_ids, new_names);
+        });
+    });
+
+</script>
+{% endblock custom_scripts %}
```
Comment (ref-humbold) [inline app/src/templates/checkboxes.html]: Directory app/src/templates should be used for generic templates. If this file was written specifically for boardgames, then it should be moved to boardgames templates

File: app/src/zosia16/settings/common.py [hunk 8]
```diff
@@ -102,6 +103,7 @@ def random_string(length=10):
     'django.contrib.sessions',
     'django.contrib.messages',
     'django.contrib.staticfiles',
+    # 'django.tables2',
```
Comment (ref-humbold) [inline app/src/zosia16/settings/common.py]: I think commented out code should be removed

File: app/src/boardgames/forms.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
+from django import forms
+from django.utils.safestring import mark_safe
+
+
+class BoardgameForm(forms.Form):
+    url = forms.CharField(
+        label=mark_safe('Find your boardgame on: <a href="https://boardgamegeek.com/">\
+            boardgamegeek.com</a> and paste here link'), max_length=200)
```
Comment (ref-humbold) [inline app/src/boardgames/forms.py]: I'm not sure whether Python will put indentation spaces into the string, so I think the feature of implicit string concatenation should be used here.

File: app/src/boardgames/views.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    if request.method == 'POST':
+        if ctx['form'].is_valid() and user_boardgames.count() < 3:
+            url = ctx['form'].cleaned_data['url']
+            validate_url(url)
+            name = get_name(url)
+            boardgame = Boardgame(name=name, user=request.user, url=url)
+            boardgame.save()
+            return redirect('my_boardgames')
+        else:
+            pass
```
Comment (ref-humbold) [inline app/src/boardgames/views.py]: You should add an error message here to be shown in toast (see e.g. views for lectures)

**Pull request merged**